schemachanger: Implement CREATE SCHEMA

Xiang-Gu opened the pull request:
We implemented CREATE SCHEMA in declarative schema changer.

We fallback when AUTHORIZATION clause is specified since we support
user-related management in declarative schema changer.

Fix #103200
Epic: [CRDB-27883](https://cockroachlabs.atlassian.net/browse/CRDB-27883)
Release note: None

Xiang-Gu committed:
schemachanger: Implement CREATE SCHEMA

We implemented CREATE SCHEMA in declarative schema changer.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/102364)
<!-- Reviewable:end -->

chengxiong-ruan commented on pkg/sql/schemachanger/scbuild/builder_state.go:
```diff
@@ -1492,6 +1509,17 @@ func (b *builderState) serializeUserDefinedTypes(queryStr string) string {
+		panic(errors.AssertionFailedf("schema name empty when resolving database prefix for a " +
+			"schema name"))
+	}
+	if schemaPrefix.CatalogName == "" {
```
`if !schemaPrefix.ExplicitCatalog` ?

chengxiong-ruan commented on pkg/sql/schemachanger/scbuild/event_log.go:
```diff
@@ -132,6 +132,25 @@ func fullyQualifiedName(b buildCtx, e scpb.Element) string {
 
+// ownerName finds the owner of the descriptor that element `e` belongs to.
+// `e` must therefore have a DescID attr.
+func ownerName(b buildCtx, e scpb.Element) string {
```
while you're here, could you also add a commit to add eventlog for CREATE FUNCTION? thanks!

chengxiong-ruan commented on pkg/sql/schemachanger/scexec/exec_immediate_mutation.go:
```diff
@@ -26,6 +27,7 @@ type immediateState struct {
 	descriptorsToDelete catalog.DescriptorIDSet
 	commentsToUpdate    []commentToUpdate
 	newDescriptors      map[descpb.ID]catalog.MutableDescriptor
+	addedNames          map[descpb.ID]descpb.NameInfo
```
instead of adding `addedNames` field here, can we use `newDescriptors` instead? I suppose we can get same information from them and creating new descriptors normally means we need to add new names except for Functions. That being said, I'm fine with your approach, but just prefer to avoid adding new fields if possible.

chengxiong-ruan commented on pkg/sql/schemachanger/scexec/scmutationexec/drop.go:
```diff
@@ -80,6 +80,16 @@ func (i *immediateVisitor) DrainDescriptorName(
 	return nil
 }
 
+func (i *immediateVisitor) AddDescriptorName(_ context.Context, op scop.AddDescriptorName) error {
```
this method should be moved to its own new file, maybe `create.go`?

chengxiong-ruan commented on pkg/sql/schemachanger/scexec/scmutationexec/function.go:
```diff
@@ -108,3 +110,27 @@ func (i *immediateVisitor) SetFunctionParamDefaultExpr(
 	return nil
 }
+
+func (i *immediateVisitor) CreateSchemaDescriptor(
```
this method can go to its own file as well, like `schema.go`

chengxiong-ruan reviewed (commented):
nice work, lgtm overall just small things.

Xiang-Gu committed:
schemachanger: event logging of creating and dropping functions

This PR supports correct event logging of creating and dropping
functions in the declarative schema changer.
 - drop function: previously implemented but it turns out it has bugs
 and `DROP FUNCTION` is not supported
 - create function: previously unimplemented

Release note: None

Xiang-Gu reviewed (commented):
I implemented event logging for `create function` (previously buggy) and `drop function`

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/102364#-:-NV_IHujCo0DQEbteA3A:b2do849)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan)

___
*[`pkg/sql/schemachanger/scbuild/builder_state.go` line 1517 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/102364#gh-1189906857:-NVUdxg_CpdjcNl9wKUP:bd2sqe2) ([raw file](https://github.com/cockroachdb/cockroach/blob/96579104bc210b38aef0f0a6377c1bf9c92bd7ad/pkg/sql/schemachanger/scbuild/builder_state.go#L1517)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

`if !schemaPrefix.ExplicitCatalog` ?
</blockquote></details>

I am just paranoid about the case (though maybe impossible) where `ExplicitCatalog` is false but `CatalogName` is not empty, in which case we should just use that `CatalogName`.
___
*[`pkg/sql/schemachanger/scexec/exec_immediate_mutation.go` line 30 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/102364#gh-1189924685:-NVUf1iX3hbzn7u1Uj3Q:b-mky1jh) ([raw file](https://github.com/cockroachdb/cockroach/blob/96579104bc210b38aef0f0a6377c1bf9c92bd7ad/pkg/sql/schemachanger/scexec/exec_immediate_mutation.go#L30)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

instead of adding `addedNames` field here, can we use `newDescriptors` instead? I suppose we can get same information from them and creating new descriptors normally means we need to add new names except for Functions. That being said, I'm fine with your approach, but just prefer to avoid adding new fields if possible.
</blockquote></details>

Maybe we can get by with just `newDescriptors` when creating new descriptors (i.e., creating a new descriptor implies adding a namespace entry, so the adding of `namespace` element is be no-op and the work is shifted to the adding path of a descriptor element), but what about deleting descriptors and future descriptor renames? 

There is a symmetry here in the following four fields:
- `descriptorsToDelete`, `drainedNames`
- `newDescriptors`, `addedNames`
___
*[`pkg/sql/schemachanger/scexec/scmutationexec/drop.go` line 83 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/102364#gh-1189927197:-NVUhr__CFX1LvKioMlx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/96579104bc210b38aef0f0a6377c1bf9c92bd7ad/pkg/sql/schemachanger/scexec/scmutationexec/drop.go#L83)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

this method should be moved to its own new file, maybe `create.go`?
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scexec/scmutationexec/function.go` line 114 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/102364#gh-1189927602:-NVUiIqS1eqddtOMIGBl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/96579104bc210b38aef0f0a6377c1bf9c92bd7ad/pkg/sql/schemachanger/scexec/scmutationexec/function.go#L114)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

this method can go to its own file as well, like `schema.go`
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

Xiang-Gu reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/102364)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan)

___
*[`pkg/sql/schemachanger/scbuild/event_log.go` line 137 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/102364#gh-1189908535:-NV_INUw3tE9eWpzC-21:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/96579104bc210b38aef0f0a6377c1bf9c92bd7ad/pkg/sql/schemachanger/scbuild/event_log.go#L137)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

while you're here, could you also add a commit to add eventlog for CREATE FUNCTION? thanks!
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

chengxiong-ruan reviewed (approved):
LGTM!

Xiang-Gu commented:
TFTR!

bors r+

Xiang-Gu commented:
bors r-

Xiang-Gu commented:
bors r+

postamar commented:
What purpose does this `SchemaName` element serve? Why couldn't `Namespace` be used?

chengxiong-ruan commented:
> What purpose does this `SchemaName` element serve? Why couldn't `Namespace` be used?

IIRC that it's used to set the name within the descriptor itself. We could have used Namespace, but then we'd need to do 2 things within one op (write namespace row + set name of the descriptor).

postamar commented:
I see. Still, the name in the descriptor and the name in the namespace table need to remain in sync at all times (not to mention the backreference in the db and whatnot). Do we have rules to ensure that these name elements all always transition in the same stage?

I'm not fundamentally opposed but I'm not convinced that the implications of this scheme have been properly thought out.

chengxiong-ruan commented:
mhm...that's a good call. I don't think we added rules to enforce the invariant :(
